Gmoccapy: Move tool table buttons to table frame, add toggle button for calculator use

Sigma1912 opened the pull request:
![tooledit classic](https://github.com/user-attachments/assets/e4caa22d-902b-4f7f-b6ce-60e3732865c4)
![tool edit material](https://github.com/user-attachments/assets/b3f20722-fa11-43db-bf63-c3a7e83329b1)

Sigma1912 commented:
Auto scroll to the bottom of the tool list when adding new tool:
![Recording](https://github.com/user-attachments/assets/bc26657a-2420-4a96-bf24-82e4e9834327)

hansu commented:
And For the save button I would suggest the check button like mentioned in https://github.com/LinuxCNC/linuxcnc/issues/2818#issuecomment-3041057883.
Yes it does a save action but also tells LinuxCNC about the changed tool table, so a check button would be mor intuitive I think.
If we are really fancy, we create a combined button.

Sigma1912 commented:
> And I would keep the four buttons left aligned - it looks a bit odd so.

I want to keep buttons that change the table to the left and buttons that dont to the right. Pulling the ones on left apart gives more room for error when using a touch screen.

hansu commented:
> Maybe we should just replace the yellow arrow on the 'back' button with the green check mark? 

Hmm you might want to have an abort option, if you accidentally removed some tools.

>One of the things that have always bugged me was the label 'Apply' which is not the same as 'Save', we might want to indicate if the table has not been saved on leaving the page.

We can add a warning dialog like when exiting from the G-code editor when having changes?

gmoccapy commented:
One question after a very very short look at the code changes:
Within tooled it widget the "new button line" is created. So the widget could be used also stand allone. But in the gmoccapy.py file the buttons from the widget are hidden and created new in gmoccapy. IMHO the buttons from the widget should send a signal and gmoccapy should just react to that signal.

Sorry, if I haven't check the code enaugh and beeping wrong here.

Norbert

hansu commented:
I only wonder why you remove the buttons inside the widget's buttonbox and create new ones. Does it not work to change only the labels to icons?

Sigma1912 commented:
How about adding a settings option to show/hide 'Select' column given that selecting multiple rows with a touchscreen only is inconvenient. If a keyboard is used the CTRL/SHIFT keys could be used to select multiple rows and the 'Select' column is not required.

hansu commented:
I mentioned resp. asked recently somewhere (didn't find now where it was) if it's the best strategy to change the buttons of the widgets like this one in the gmoccapy code and not in the widget itself.
I think it would be better for maintainability if this is done in the widget's code. Sure, this will affect other GUIs like Gscreen which use these widgets, but maybe they benefit also from these changes. @c-morley what do you think?

c-morley commented:
If buttons makes sense for general use, (I didn't read back enough to see what buttons you are referring to) then I would think adding them to the widget would be better.
Similarly you could add a function that would be used to add a button to the existing button box, so a third party could add whatever buttons they want.

If you must change all the buttons for some reason, then adding a function or property to hide/delete the button box and (possibly) add a new button box with new specified buttons would seem reasonable.

hansu commented:
>If buttons makes sense for general use, (I didn't read back enough to see what buttons you are referring to) then I would think adding them to the widget would be better.

It's more about changing the existing buttons to have an icon instead of just a label and change the size and maybe the order.

Sigma1912 commented:
> Sure we can add a setting to hide the select column. But I don't really get the point why - Is it mostly to avoid confusion? Or space reasons?

It's to avoid confusion. The only use case for the Select column is to select multiple tools with a touch screen w/o a keyboard.

c-morley commented:
> > If buttons makes sense for general use, (I didn't read back enough to see what buttons you are referring to) then I would think adding them to the widget would be better.
> 
> It's more about changing the existing buttons to have an icon instead of just a label and change the size and maybe the order.

Can that not be controlled with css styles?

Sigma1912 commented:
Modified and rebased as requested

<img width="892" height="519" alt="tooltable" src="https://github.com/user-attachments/assets/b3c09d9c-e523-4345-baf2-608f7fe1234c" />

hansu commented:
Nice, that was quick. Thanks!

hansu commented:
>I wanted to ask if you are using the buttons from the tooledit widget?

Yes they are used and re-arranged. But the tooltips seem to got lost. Maybe even in an earlier commit.

>If I understand correctly, there are unused functions left in gmoccapy.py:
def on_btn_delete_tool_clicked(self, widget, data=None):
def on_btn_add_tool_clicked(self, widget, data=None):
def on_btn_reload_tooltable_clicked(self, widget, data=None):
def on_btn_save_tool_changes_clicked(self, widget, data=None):


Yes it seems they are unused. E.g. this one is not called. It's currently possible to delete the current tool which this function disallows:


```
    def on_btn_delete_tool_clicked(self, widget, data=None):
        act_tool = self.stat.tool_in_spindle
        if act_tool == self.widgets.tooledit1.get_selected_tool():
            message = _("You are trying to delete the tool mounted in the spindle\n")
            message += _("This is not allowed, please change tool prior to delete it")
            self.dialogs.warning_dialog(self, _("Warning Tool can not be deleted!"), message)
            return

```

Sigma1912 commented:
See https://github.com/LinuxCNC/linuxcnc/pull/3706

Thanks for testing.

zz912 commented:
> The check could be improved to really avoid redundance like "If tool info is shown --> hide tool information in table". But it's a bit to much effort for just avoidung that redundance.

I love details. This should be nice detail. Can I ask Sigma for it? Please.
